### PR TITLE
[BEAM-10305] Let InMemoryBagUserStateFactory only use a single cache token

### PR DESCRIPTION
**Pull request by mxm**
When the state cache is enabled in the Python SDK, the batch mode of the Flink
Runner currently only allows a single user state cell because a new cache token
is generated for each state cell; the caching code in the Python SDK Harness
only supports one cache token per user state handler.

Theoretically multiple cache tokens would work but would just be adding to the
payload. We should make sure to just send a single cache token in batch
mode (which is already the case in streaming)

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Dataflow | Flink | Samza | Spark
--- | --- | --- | --- | --- | ---
Go | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Go_VR_Flink/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Go_VR_Spark/lastCompletedBuild/)
Java | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow_Java11/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink_Java11/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Spark_Batch/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_SparkStructuredStreaming/lastCompletedBuild/)
Python | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python2/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python35/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python36/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python37/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Py_VR_Dataflow_V2/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Python2_PVR_Flink_Cron/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python35_VR_Flink/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python_VR_Spark/lastCompletedBuild/)
XLang | --- | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_XVR_Flink/lastCompletedBuild/) | --- | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_XVR_Spark/lastCompletedBuild/)

Pre-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

--- |Java | Python | Go | Website
--- | --- | --- | --- | ---
Non-portable | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Java_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Python_Cron/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PreCommit_PythonLint_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Go_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Website_Cron/lastCompletedBuild/)
Portable | --- | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Portable_Python_Cron/lastCompletedBuild/) | --- | ---

See [.test-infra/jenkins/README](https://github.com/apache/beam/blob/master/.test-infra/jenkins/README.md) for trigger phrase, status and link of all Jenkins jobs.

### CODE REVIEW

**Comment by mxm** [comment]
Run Java PreCommit

File: runners/java-fn-execution/src/main/java/org/apache/beam/runners/fnexecution/state/InMemoryBagUserStateFactory.java [hunk 1]
```diff
@@ -41,9 +41,12 @@
 public class InMemoryBagUserStateFactory<K, V, W extends BoundedWindow>
```
Comment (lukecwik) [inline runners/java-fn-execution/src/main/java/org/apache/beam/runners/fnexecution/state/InMemoryBagUserStateFactory.java:41]: You should update the comment related to error handling to make sure that users discard the factory every time they want a new cache token.

**Comment by y1chi** [comment]
IIUC, this changes all state handlers to use same cache token value, but doesn't reduce the number of user state cache token in process bundle request to 1?

**Comment by mxm** [comment]
> IIUC, this changes all state handlers to use same cache token value, but doesn't reduce the number of user state cache token in process bundle request to 1?

Yes, it reduces the number of user state cache tokens for process bundle requests per process bundle descriptor to 1.

File: runners/java-fn-execution/src/main/java/org/apache/beam/runners/fnexecution/state/InMemoryBagUserStateFactory.java (hunk 1 above)
Comment (mxm) [reply runners/java-fn-execution/src/main/java/org/apache/beam/runners/fnexecution/state/InMemoryBagUserStateFactory.java:44]: Sure, good idea to clarify that.

**Comment by lukecwik** [comment]
> > IIUC, this changes all state handlers to use same cache token value, but doesn't reduce the number of user state cache token in process bundle request to 1?
> 
> Yes, it reduces the number of user state cache tokens for process bundle requests per process bundle descriptor to 1.

I think the logic in `ByteStringStateRequestHandlerToBagUserStateHandlerFactoryAdapter#getCacheTokens` https://github.com/apache/beam/blob/f98104a22b69972744a13378e17af5f2361fbb3e/runners/java-fn-execution/src/main/java/org/apache/beam/runners/fnexecution/state/StateRequestHandlers.java#L590 is going to add N copies of the same cache token to the `ProcessBundleRequest`.

File: runners/java-fn-execution/src/main/java/org/apache/beam/runners/fnexecution/state/InMemoryBagUserStateFactory.java [hunk 2]
```diff
@@ -37,13 +37,19 @@
...
  */
 public class InMemoryBagUserStateFactory<K, V, W extends BoundedWindow>
     implements StateRequestHandlers.BagUserStateHandlerFactory<K, V, W> {
 
+  private final ByteString cacheToken;
+
   private List<InMemorySingleKeyBagState> handlers;
 
   public InMemoryBagUserStateFactory() {
+    cacheToken = ByteString.copyFrom(UUID.randomUUID().toString().getBytes(Charsets.UTF_8));
```
Comment (lukecwik) [inline runners/java-fn-execution/src/main/java/org/apache/beam/runners/fnexecution/state/InMemoryBagUserStateFactory.java:52]: I think we'll want to move the cache token generation up higher to the `ByteStringStateRequestHandlerToBagUserStateHandlerFactoryAdapter`. This comment added for this class should be moved to the ByteStringStateRequestHandlerToBagUserStateHandlerFactoryAdapter

We should also test that it only adds a single cache token to the request.

Comment (mxm) [reply runners/java-fn-execution/src/main/java/org/apache/beam/runners/fnexecution/state/InMemoryBagUserStateFactory.java:52]: Could you add a reason for moving this up the stack? State handlers should implement their own cache token handling.

Comment (mxm) [reply runners/java-fn-execution/src/main/java/org/apache/beam/runners/fnexecution/state/InMemoryBagUserStateFactory.java:52]: I'll have another look later.

Comment (lukecwik) [reply runners/java-fn-execution/src/main/java/org/apache/beam/runners/fnexecution/state/InMemoryBagUserStateFactory.java:52]: `ByteStringStateRequestHandlerToBagUserStateHandlerFactoryAdapter#getCacheTokens` is going to add N copies of the cache token to the `ProcessBundleRequest`

For user state, it makes sense to have a caching handler that does the delegation to other handlers. This caching handler is the one that should be responsible for supplying the single cache token.

Comment (mxm) [reply runners/java-fn-execution/src/main/java/org/apache/beam/runners/fnexecution/state/InMemoryBagUserStateFactory.java:52]: I see, makes sense. Let me update the PR.

**Comment by mxm** [comment]
Thanks for pointing out the duplication issue. The new version should properly fix it. Please have another look.

File: runners/java-fn-execution/src/test/java/org/apache/beam/runners/fnexecution/state/StateRequestHandlersTest.java [hunk 3]
```diff
@@ -59,4 +68,46 @@ public void testDelegatingStateHandlerThrowsWhenNotFound() throws Exception {
...
+        stateRequestHandler.getCacheTokens();
+    assertThat(Iterables.size(cacheTokens), is(1));
+
+    BeamFnApi.ProcessBundleRequest.CacheToken cacheToken = Iterables.getOnlyElement(cacheTokens);
+    assertThat(
+        cacheToken.getUserState(),
+        is(BeamFnApi.ProcessBundleRequest.CacheToken.UserState.getDefaultInstance()));
+    assertThat(cacheToken.getToken(), is(notNullValue()));
+
+    inMemoryBagUserStateFactory.forUserState(
```
Comment (mxm) [inline runners/java-fn-execution/src/test/java/org/apache/beam/runners/fnexecution/state/StateRequestHandlersTest.java:91]: Looks like this should be using the StateRequestHandler instead, will update.

**Comment by mxm** [comment]
Run Java PreCommit

**Comment by mxm** [comment]
Run Java PreCommit

**Comment by mxm** [comment]
Unrelated test failures in https://ci-beam.apache.org/job/beam_PreCommit_Java_Phrase/2367/

```
14:39:21 FAILURE: Build completed with 2 failures.
14:39:21 
14:39:21 1: Task failed with an exception.
14:39:21 -----------
14:39:21 * What went wrong:
14:39:21 Execution failed for task ':sdks:java:io:jdbc:test'.
14:39:21 > There were failing tests. See the report at: file:///home/jenkins/jenkins-slave/workspace/beam_PreCommit_Java_Phrase/src/sdks/java/io/jdbc/build/reports/tests/test/index.html
14:39:21 
14:39:21 * Try:
14:39:21 Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output. Run with --scan to get full insights.
14:39:21 ==============================================================================
14:39:21 
14:39:21 2: Task failed with an exception.
14:39:21 -----------
14:39:21 * What went wrong:
14:39:21 Execution failed for task ':sdks:java:io:rabbitmq:test'.
14:39:21 > Process 'Gradle Test Executor 124' finished with non-zero exit value 143
14:39:21   This problem might be caused by incorrect test process configuration.
14:39:21   Please refer to the test execution section in the User Manual at https://docs.gradle.org/5.2.1/userguide/java_testing.html#sec:test_execution
14:39:21 
```

**Pull request merged**